prefect_azure/container_instance.py

```diff
@@ -0,0 +1,694 @@
+
+    Run a command and stream the container's output to the local terminal.
+    ```python
+    ContainerInstanceJob(
+        command=["echo", "hello world"],
+        stream_output=True,
+    )
+    ```
+
+    Run a command with a specific image
+    ```python
+    ContainerInstanceJob(command=["echo", "hello world"], image="alpine:latest")
+    ```
+
+    Run a task with custom memory and CPU requirements
+    ```python
+    ContainerInstanceJob(command=["echo", "hello world"], memory=1.0, cpu=1.0)
+    ```
+
+    Run a task with custom memory and CPU requirements
+    ```python
+    ContainerInstanceJob(command=["echo", "hello world"], memory=1.0, cpu=1.0)
+    ```
+
+    Run a task with custom memory, CPU, and GPU requirements
+    ```python
+    ContainerInstanceJob(command=["echo", "hello world"], memory=1.0, cpu=1.0,
+    gpu_count=1, gpu_sku="V100")
+    ```
+
+    Run a task with custom environment variables
+    ```python
+    ContainerInstanceJob(command=["echo", "hello $PLANET"], env={"PLANET": "earth"})
+    ```
+"""
+import datetime
+import json
+import sys
+import time
+import uuid
+from enum import Enum
+from typing import Dict, List, Optional
+
+import dateutil.parser
+import prefect.infrastructure.docker
+from anyio.abc import TaskStatus
+from azure.core.credentials import TokenCredential
+from azure.core.polling import LROPoller
+from azure.identity import ClientSecretCredential
+from azure.mgmt.containerinstance import ContainerInstanceManagementClient
+from azure.mgmt.containerinstance.models import (
+    Container,
+    ContainerGroup,
+    ContainerGroupRestartPolicy,
+    ContainerGroupSubnetId,
+    EnvironmentVariable,
+    GpuResource,
+    ImageRegistryCredential,
+    OperatingSystemTypes,
+    ResourceRequests,
+    ResourceRequirements,
+)
+from azure.mgmt.resource import ResourceManagementClient
+from prefect.docker import get_prefect_image_name
+from prefect.infrastructure.base import Infrastructure, InfrastructureResult
+from prefect.utilities.asyncutils import run_sync_in_worker_thread, sync_compatible
+from pydantic import Field, SecretStr
+from typing_extensions import Literal
+
+from prefect_azure.credentials import ContainerInstanceCredentials
+
+ACI_DEFAULT_CPU = 1.0
+ACI_DEFAULT_MEMORY = 1.0
+ACI_DEFAULT_GPU = 0.0
+DEFAULT_CONTAINER_ENTRYPOINT = "/opt/prefect/entrypoint.sh"
+
+
+class ContainerGroupProvisioningState(str, Enum):
+    """
+    Terminal provisioning states for ACI container groups. Per the Azure docs,
+    the states in this Enum are the only ones that can be relied on as dependencies.
+    """
+
+    SUCCEEDED = "Succeeded"
+    FAILED = "Failed"
+
+
+class ContainerRunState(str, Enum):
+    """
+    Terminal run states for ACI containers.
+    """
+
+    RUNNING = "Running"
+    TERMINATED = "Terminated"
+
+
+class ContainerInstanceJobResult(InfrastructureResult):
+    """
+    The result of an `ContainerInstanceJob` run.
+    """
+
+
+class ContainerInstanceJob(Infrastructure):
```

Comment:
I think the name of this class should be `AzureContainerInstanceBlock` so that it matches the block name and the naming convention of other blocks in this collection.
```suggestion
class AzureContainerInstanceJob(Infrastructure):
```